### PR TITLE
Clean-up SZA calculation

### PR DESCRIPTION
**Pull request by claresinger**
Update zenith angle calculation.
- Uses equations from Tapio's book.
- Removes part that calculates orbital parameters, and just requires them as direct input.
- Add azimuth calculation for land model.
- Insolation takes zenith angle as input to remove redundancies.
- Update docs.
- Removed tests, need to add new ones back later maybe.

### CODE REVIEW

File: src/ZenithAngleCalc.jl [hunk 1]
```diff
@@ -204,103 +28,88 @@ function instantaneous_zenith_angle(date::DateTime,
...
-    GMSTrad = GMST(date, timezone)
-    d = earth_sun_dist(ecc, TA)
 
-    # declination, radians
+    # declination, radians (3.16)
     δ = mod(asin(sin(γ) * sin(TL)), 2*π)
-    # right acension, radians
-    RA = mod(atan(cos(γ) * sin(TL) / cos(TL)), 2*π)
 
     # hour angle, radians
```
Comment (szy21) [inline src/ZenithAngleCalc.jl:53]:
```suggestion
    # hour angle, radians (3.17)
```
(It's not exactly the same but I think it's still helpful to add the equation number.)

**Comment by szy21** [comment]
Thanks Clare! I take a look at the calculations in `InsolationCalc.jl` and `ZenithAngleCalc.jl`, and they look good to me. Two general questions:

1. Do we need to add FT before the numbers in the functions?
2. Should we move OrbitalParameters to CLIMAParameters?

**Comment by claresinger** [comment]
Thanks @szy21! To answer your questions: 
1. I'm not sure. This was just something I'd seen in other Julia code, but we should ask an expert!
2. Probably. @charleskawczynski can you help with this?

File: src/ZenithAngleCalc.jl [hunk 2]
```diff
@@ -204,103 +28,88 @@ function instantaneous_zenith_angle(date::DateTime,
...
+    e = eccentricity
 
-    TA = true_anomaly(date)
+    # mean anomaly at vernal equinox, radians (3.11)
+    β = (1-e^2)^(0.5)
+    MAve = -ϖ + (e + e^3/4)*(1+β)*sin(ϖ) - 0.5*e^2*(0.5+β)*sin(2*ϖ) + 0.25*e^3*(1/3+β)*sin(3*ϖ)
+
+    # mean anomaly, radians (3.10)
+    julian_day_abs = datetime2julian(date)
+    julian_day_ref = 2451545.0 # Jan 1, 2000 at 12hr
```
Comment (tapios) [inline src/ZenithAngleCalc.jl:39]: We should make this a parameter (called epoch), so it is easy to adjust for other planets.

File: src/ZenithAngleCalc.jl [hunk 3]
```diff
@@ -204,103 +28,88 @@ function instantaneous_zenith_angle(date::DateTime,
...
+    julian_day_ve = julian_day_ref + 79.0 # vernal equinox 2000
```
Comment (tapios) [inline src/ZenithAngleCalc.jl:40]: Likewise, make this a parameter, along with other orbital parameters.

File: src/ZenithAngleCalc.jl [hunk 4]
```diff
@@ -204,103 +28,88 @@ function instantaneous_zenith_angle(date::DateTime,
...
     ecc = eccentricity
-    ϕ = deg2rad(latitude)
 
-    # calculate mean anomaly at vernal equinox and mean anomaly
-    β = (1 - ecc^2)^0.5
-    MA_VE = mod(-ϖ + (ecc + ecc^3/4)*(1+β)*sin(ϖ), 2*π)
-    MA = mod(2*π*days_since_equinox / (year_anom() / day_length()) + MA_VE, 2*π)
+    # mean anomaly at vernal equinox, radians (3.11)
+    β = (1-e^2)^(0.5)
+    MAve = -ϖ + (e + e^3/4)*(1+β)*sin(ϖ) - 0.5*e^2*(0.5+β)*sin(2*ϖ) + 0.25*e^3*(1/3+β)*sin(3*ϖ)
```
Comment (tapios) [inline src/ZenithAngleCalc.jl:97]: This appears for the second time. Better to make it a function to only have the formula stated once.

File: src/ZenithAngleCalc.jl [hunk 5]
```diff
@@ -204,103 +28,88 @@ function instantaneous_zenith_angle(date::DateTime,
...
 
-    # solar longitude and true anomaly
-    TA = mod(MA + (2*ecc - ecc^3/4)*sin(MA), 2*π)
-    TL = mod(TA + ϖ, 2*π)
+    # mean anomaly, radians (3.10)
+    julian_day_abs = datetime2julian(date)
+    julian_day_ref = 2451545.0 # Jan 1, 2000 at 12hr
+    julian_day_ve = julian_day_ref + 79.0 # vernal equinox 2000
```
Comment (tapios) [inline src/ZenithAngleCalc.jl:102]: As above, make the epoch and time of vernal equinox in year of epoch parameters.

File: src/ZenithAngleCalc.jl [hunk 6]
```diff
@@ -204,103 +28,88 @@ function instantaneous_zenith_angle(date::DateTime,
...
-    # radius earth-sun distance, AU and m
-    d = earth_sun_dist(ecc, TA)
+    # true anomaly, radians (3.8)
+    TA = MA + (2*e - e^3/4)*sin(MA) + (5/4)*e^2*sin(2*MA) + (13/12)*e^3*sin(3*MA)
 
-    # declination, radians
+    # true longitude, radians (3.9)
+    TL = mod(TA + ϖ, 2*π)
+
+    # declination, radians (3.16)
```
Comment (tapios) [inline src/ZenithAngleCalc.jl:112]: Make this entire chunk of code a function, or several functions, to not have the same quantities defined twice.

File: src/ZenithAngleCalc.jl [hunk 7]
```diff
@@ -310,12 +119,15 @@ function daily_zenith_angle(days_since_equinox::I,
     elseif T <= -1
         ηd = 0.0
     else
-        ηd = acos(-1 * T)
+        ηd = acos(-1*T)
     end
     
     # daily averaged zenith angle
     szabar = acos((1/π)*(ηd*sin(ϕ)*sin(δ) + cos(ϕ)*cos(δ)*sin(ηd)))
```
Comment (tapios) [inline src/ZenithAngleCalc.jl:126]: Maybe a name like `daily_sza` is more intuitive?

**Review by tapios — APPROVED** [review]
Thank you for putting this together, @claresinger. It all looks good. Please just make sure you define functions for code segments (mean anomaly calculation etc.) that are being reused. A few other points to consider:

- It would also be good to add a few simple tests, e.g., symmetry of insolation at equinoxes, and invariance of zonal-mean insolation under rotations of perihelion. 
- Please consult with @kpamnany about the date format. We want to make this consistent with whatever we will be using in the model globally. (Julian time is fine with me.)

File: src/ZenithAngleCalc.jl (hunk 2 above)
Comment (charleskawczynski) [inline src/ZenithAngleCalc.jl:39]:
```suggestion
    β = (1-e^2)^FT(0.5)
    MAve = -ϖ + (e + e^3/4)*(1+β)*sin(ϖ) - FT(0.5)*e^2*(FT(0.5)+β)*sin(2*ϖ) + FT(0.25)*e^3*(FT(1/3)+β)*sin(3*ϖ)

    # mean anomaly, radians (3.10)
    julian_day_abs = datetime2julian(date)
    julian_day_ref = FT(2451545.0) # Jan 1, 2000 at 12hr
    julian_day_ve = julian_day_ref + FT(79.0) # vernal equinox 2000
```

File: src/ZenithAngleCalc.jl [hunk 8]
```diff
@@ -204,103 +28,88 @@ function instantaneous_zenith_angle(date::DateTime,
...
+
+    # mean anomaly, radians (3.10)
+    julian_day_abs = datetime2julian(date)
+    julian_day_ref = 2451545.0 # Jan 1, 2000 at 12hr
+    julian_day_ve = julian_day_ref + 79.0 # vernal equinox 2000
+    t_since_vernal_equinox = mod(julian_day_abs - julian_day_ve, year_anom())
+    MA = 2π * t_since_vernal_equinox / year_anom() + MAve
+
+    # true anomaly, radians (3.8)
+    TA = MA + (2*e - e^3/4)*sin(MA) + (5/4)*e^2*sin(2*MA) + (13/12)*e^3*sin(3*MA)
```
Comment (charleskawczynski) [inline src/ZenithAngleCalc.jl:45]:
```suggestion
    TA = MA + (2*e - e^3/4)*sin(MA) + FT(5/4)*e^2*sin(2*MA) + FT(13/12)*e^3*sin(3*MA)
```

File: src/ZenithAngleCalc.jl [hunk 9]
```diff
@@ -204,103 +28,88 @@ function instantaneous_zenith_angle(date::DateTime,
...
+    # declination, radians (3.16)
     δ = mod(asin(sin(γ) * sin(TL)), 2*π)
-    # right acension, radians
-    RA = mod(atan(cos(γ) * sin(TL) / cos(TL)), 2*π)
 
     # hour angle, radians
-    LMST = GMSTrad + λ
-    η = mod(LMST - RA, 2*π)
+    julian_day_abs = datetime2julian(date)
+    η_UTC = 2π * (mod(julian_day_abs, 1.0) - 0.5)
```
Comment (charleskawczynski) [inline src/ZenithAngleCalc.jl:55]:
```suggestion
    η_UTC = 2π * (mod(julian_day_abs, 1) - FT(0.5))
```

File: src/ZenithAngleCalc.jl [hunk 10]
```diff
@@ -204,103 +28,88 @@ function instantaneous_zenith_angle(date::DateTime,
...
-                       latitude::FT) where {FT <: Real}
-
-returns the daily averaged zenith angle and earth-sun distance
-at a particular latitude on the given date
-"""
-function daily_zenith_angle(date::DateTime,
-                            latitude::FT) where {FT <: Real}
-    ϕ = deg2rad(latitude)
+    # earth-sun distance, m (3.1)
+    d = astro_unit() * (1.0 - e^2) / (1.0 + e*cos(TA))
```
Comment (charleskawczynski) [inline src/ZenithAngleCalc.jl:70]:
```suggestion
    d = astro_unit() * (1 - e^2) / (1 + e*cos(TA))
```

File: src/ZenithAngleCalc.jl (hunk 4 above)
Comment (charleskawczynski) [inline src/ZenithAngleCalc.jl:97]:
```suggestion
    β = (1-e^2)^FT(0.5)
    MAve = -ϖ + (e + e^3/4)*(1+β)*sin(ϖ) - FT(0.5)*e^2*(FT(0.5)+β)*sin(2*ϖ) + FT(0.25)*e^3*(FT(1/3)+β)*sin(3*ϖ)
```

File: src/ZenithAngleCalc.jl (hunk 5 above)
Comment (charleskawczynski) [inline src/ZenithAngleCalc.jl:102]:
```suggestion
    julian_day_ref = FT(2451545.0) # Jan 1, 2000 at 12hr
    julian_day_ve = julian_day_ref + FT(79.0) # vernal equinox 2000
```

File: src/ZenithAngleCalc.jl [hunk 11]
```diff
@@ -204,103 +28,88 @@ function instantaneous_zenith_angle(date::DateTime,
...
+    julian_day_abs = datetime2julian(date)
+    julian_day_ref = 2451545.0 # Jan 1, 2000 at 12hr
+    julian_day_ve = julian_day_ref + 79.0 # vernal equinox 2000
+    t_since_vernal_equinox = mod(julian_day_abs - julian_day_ve, year_anom())
+    MA = 2π * t_since_vernal_equinox / year_anom() + MAve
 
-    # radius earth-sun distance, AU and m
-    d = earth_sun_dist(ecc, TA)
+    # true anomaly, radians (3.8)
+    TA = MA + (2*e - e^3/4)*sin(MA) + (5/4)*e^2*sin(2*MA) + (13/12)*e^3*sin(3*MA)
```
Comment (charleskawczynski) [inline src/ZenithAngleCalc.jl:107]:
```suggestion
    TA = MA + (2*e - e^3/4)*sin(MA) + FT(5/4)*e^2*sin(2*MA) + FT(13/12)*e^3*sin(3*MA)
```

File: src/ZenithAngleCalc.jl [hunk 12]
```diff
@@ -310,12 +119,15 @@ function daily_zenith_angle(days_since_equinox::I,
...
     else
-        ηd = acos(-1 * T)
+        ηd = acos(-1*T)
     end
     
     # daily averaged zenith angle
     szabar = acos((1/π)*(ηd*sin(ϕ)*sin(δ) + cos(ϕ)*cos(δ)*sin(ηd)))
 
+    # earth-sun distance, m (3.1)
+    d = astro_unit() * (1.0 - e^2) / (1.0 + e*cos(TA))
```
Comment (charleskawczynski) [inline src/ZenithAngleCalc.jl:129]:
```suggestion
    d = astro_unit() * (1 - e^2) / (1 + e*cos(TA))
```

**Comment by tapios** [comment]
Some good tests would compute globally averaged insolation (compute it varies latitudes and integrate numerically) and make sure that is close to the expected values given TSI.

File: src/ZenithAngleCalc.jl [hunk 13]
```diff
@@ -1,322 +1,116 @@
...
+    # mean anomaly given mean anomaly at vernal equinox (3.10)
+    time = datetime2julian(date)*day_length
+    M_v = mean_anomaly_vernal_equinox(ϖ, e)
+    MA = mod(2π * (time - time_v) / Ya + M_v, 2π)
 
-    # hour angle, radians
-    LMST = mod(GMSTrad + λ, 2*π)
-    η = mod(LMST - RA, 2*π)
+    # true anomaly, radians (3.8)
+    TA = mod(MA + (2*e - e^FT(3/4))*sin(MA) + FT(5/4)*e^2*sin(2*MA) + FT(13/12)*e^3*sin(3*MA), 2π)
```
Comment (tapios) [inline src/ZenithAngleCalc.jl:25]:
```suggestion
    TA = mod(MA + (2*e - e^FT(3)/4)*sin(MA) + FT(5/4)*e^2*sin(2*MA) + FT(13/12)*e^3*sin(3*MA), 2π)
```
(That's may be why you got the error for the rotation of perihelion.) You should also encapsulate constants here and in other formulas in `FT(...)`, and use things such as `0.25 *...` and `.../4` consistently.

Comment (claresinger) [reply src/ZenithAngleCalc.jl:25]: Yes! Problem solved. Thank you for spotting that typo.

Comment (claresinger) [reply src/ZenithAngleCalc.jl]: Looks good :) 
insol_example2b

File: src/InsolationCalc.jl [hunk 14]
```diff
@@ -1,4 +1,4 @@
-export insolation
+export insolation, solar_flux_and_mu
```
Comment (tapios) [inline src/InsolationCalc.jl:1]: `solar_flux_and_cos_sza` for more verbosity?

**Comment by szy21** [comment]
I talked with Clare and am posting it here in case anyone has suggestions. Solar flux is just a function of time, not a function of latitude or longitude, so we don't need to calculate it at each point. I'm not sure whether it's more efficient to calculate it for all points or to just calculate it once at one timestep and store it?

We don't have to make any change now. I am just looking for suggestions.

**Comment by skandalaCLIMA** [comment]
The interface looks good to me.

**Review by skandalaCLIMA — APPROVED** [review]
The interface looks good to me.

**Pull request merged**